File: app/src/main/java/de/westnordost/streetcomplete/location/LocationStateButton.java
package de.westnordost.streetcomplete.location;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.drawable.Animatable;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;

import java.util.Arrays;

import de.westnordost.streetcomplete.R;

/**
 * An image button which shows the current location state
 */
public class LocationStateButton extends androidx.appcompat.widget.AppCompatImageButton
{
	// must be defined in the same order as the LocationState enum (but minus the first)
	private static final int[] STATES = {
			R.attr.state_allowed,
			R.attr.state_enabled,
			R.attr.state_searching,
			R.attr.state_updating,
	};

	private LocationState state;

	private ColorStateList tint;

	public LocationStateButton(Context context)
	{
		this(context, null);
	}

	public LocationStateButton(Context context, AttributeSet attrs)
	{
		this(context, attrs, 0);
	}

	public LocationStateButton(Context context, AttributeSet attrs, int defStyle)
	{
		super(context, attrs, defStyle);

		TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.LocationStateButton);
		state = determineStateFrom(a);
		tint = a.getColorStateList(R.styleable.LocationStateButton_tint);
		a.recycle();
	}

	private LocationState determineStateFrom(TypedArray a)
	{
		if (a.getBoolean(R.styleable.LocationStateButton_state_updating, false)) return LocationState.UPDATING;
		if (a.getBoolean(R.styleable.LocationStateButton_state_searching, false)) return LocationState.SEARCHING;
		if (a.getBoolean(R.styleable.LocationStateButton_state_enabled, false)) return LocationState.ENABLED;
		if (a.getBoolean(R.styleable.LocationStateButton_state_allowed, false)) return LocationState.ALLOWED;
		return LocationState.DENIED;
	}

	@Override protected void drawableStateChanged()
	{
		super.drawableStateChanged();
		// autostart
		if(getDrawable().getCurrent() instanceof Animatable)
		{
			Animatable animatable = (Animatable) getDrawable().getCurrent();
			if(!animatable.isRunning()) animatable.start();
		}
		if (tint != null && tint.isStateful())
		{
			setColorFilter(tint.getColorForState(getDrawableState(), 0));
		}
	}

	public void setState(LocationState state)
	{
		if(state == this.state) return;
		this.state = state;
		refreshDrawableState();
	}

	public LocationState getState()
	{
		if(state == null) return LocationState.DENIED;
		return state;
	}

	@Override
	public int[] onCreateDrawableState(int extraSpace)
	{
		int additionalLength = STATES.length + 1;
		final int[] drawableState = super.onCreateDrawableState(extraSpace + additionalLength);
		int arrPos = getState().ordinal();
		int[] additionalArray = Arrays.copyOf(Arrays.copyOf(STATES, arrPos), additionalLength);
		mergeDrawableStates(drawableState, additionalArray);
		return drawableState;
	}

	@Override
	public Parcelable onSaveInstanceState()
	{
		Parcelable superState = super.onSaveInstanceState();
		SavedState ss = new SavedState(superState);
		ss.state = getState();
		ss.activated = isActivated();
		return ss;
	}

	@Override
	public void onRestoreInstanceState(Parcelable s)
	{
		SavedState ss = (SavedState) s;
		super.onRestoreInstanceState(ss.getSuperState());
		this.state = ss.state;
		setActivated(ss.activated);
		requestLayout();
	}

	static class SavedState extends BaseSavedState
	{
		LocationState state;
		boolean activated;

		SavedState(Parcelable superState) {
			super(superState);
		}

		private SavedState(Parcel in)
		{
			super(in);
			state = LocationState.valueOf(in.readString());
			activated = in.readInt() == 1;
		}

		@Override public void writeToParcel(Parcel out, int flags) {
			super.writeToParcel(out, flags);
			out.writeString(state.name());
			out.writeInt(activated ? 1 : 0);
		}

		public static final Parcelable.Creator<SavedState> CREATOR
				= new Parcelable.Creator<SavedState>() {
			public SavedState createFromParcel(Parcel in) {
				return new SavedState(in);
			}

			public SavedState[] newArray(int size) {
				return new SavedState[size];
			}
		};
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/location/LocationStateButton.kt
package de.westnordost.streetcomplete.location

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.TypedArray
import android.graphics.drawable.Animatable
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import androidx.annotation.Keep
import androidx.appcompat.widget.AppCompatImageButton
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.location.LocationState.*

/**
 * An image button which shows the current location state
 */
class LocationStateButton @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : AppCompatImageButton(context, attrs, defStyle) {

    var state: LocationState
    get() = _state ?: DENIED
    set(value) { _state = value }

    // this is necessary because state is accessed before it is initialized (in constructor of super)
    private var _state: LocationState? = null
    set(value) {
        if (field != value) {
            field = value
            refreshDrawableState()
        }
    }

    private val tint: ColorStateList?

    var isNavigation: Boolean = false
    set(value) {
        if (field != value) {
            field = value
            refreshDrawableState()
        }
    }


    init {
        val a = context.obtainStyledAttributes(attrs, R.styleable.LocationStateButton)
        state = determineStateFrom(a)
        tint = a.getColorStateList(R.styleable.LocationStateButton_tint)
        isNavigation = a.getBoolean(R.styleable.LocationStateButton_is_navigation, false)

        a.recycle()
    }

    private fun determineStateFrom(a: TypedArray): LocationState = when {
        a.getBoolean(R.styleable.LocationStateButton_state_updating,false) ->  UPDATING
        a.getBoolean(R.styleable.LocationStateButton_state_searching,false) -> SEARCHING
        a.getBoolean(R.styleable.LocationStateButton_state_enabled,false) ->   ENABLED
        a.getBoolean(R.styleable.LocationStateButton_state_allowed,false) ->   ALLOWED
        else -> DENIED
    }

    override fun drawableStateChanged() {
        super.drawableStateChanged()
        // autostart
        val current = drawable.current
        if (current is Animatable) {
            if (!current.isRunning) current.start()
        }
        if (tint != null && tint.isStateful) {
            setColorFilter(tint.getColorForState(drawableState, 0))
        }
    }

    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        val attributes = ArrayList<Int>()
        attributes += state.styleableAttributes
        if (isNavigation) attributes += R.attr.is_navigation

        val drawableState = super.onCreateDrawableState(extraSpace + attributes.size)

        View.mergeDrawableStates(drawableState, attributes.toIntArray())
        return drawableState
    }

    public override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        val ss = SavedState(superState)
        ss.state = state
        ss.activated = isActivated
        ss.navigation = isNavigation
        return ss
    }

    public override fun onRestoreInstanceState(s: Parcelable) {
        val ss = s as SavedState
        super.onRestoreInstanceState(ss.superState)
        state = ss.state
        isActivated = ss.activated
        isNavigation = ss.navigation
        requestLayout()
    }

    internal class SavedState : BaseSavedState {
        var state: LocationState = DENIED
        var activated = false
        var navigation = false

        constructor(superState: Parcelable?) : super(superState)
        constructor(parcel: Parcel) : super(parcel) {
            state = LocationState.valueOf(parcel.readString()!!)
            activated = parcel.readInt() == 1
            navigation = parcel.readInt() == 1
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeString(state.name)
            out.writeInt(if (activated) 1 else 0)
            out.writeInt(if (navigation) 1 else 0)
        }

        companion object {
            @JvmField @Keep
            val CREATOR = object : Parcelable.Creator<SavedState> {
                override fun createFromParcel(parcel: Parcel) = SavedState(parcel)
                override fun newArray(size: Int) = arrayOfNulls<SavedState>(size)
            }
        }
    }
}

private val LocationState.styleableAttributes: List<Int> get() =
    listOf(
        R.attr.state_allowed,
        R.attr.state_enabled,
        R.attr.state_searching,
        R.attr.state_updating
    ).subList(0, ordinal)
